hk_adc/hk_adc.c

```diff
@@ -0,0 +1,504 @@
+/*
+ * HK-ADC functions
+ *
+ * Copyright © 2020 Theodor Stana, Sonal Shrivastava
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the “Software”), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "../firmware/drivers/mss_gpio/mss_gpio.h"
+#include "../firmware/drivers/mss_i2c/mss_i2c.h"
+#include "hk_adc.h"
+
+
+#define TX_LENGTH                       (1u)
+#define RX_LENGTH                       (2u)
+#define DUMMY_SERIAL_ADDR               (0xFF)
+#define NUM_SAMPLES_TO_AVG              (10u)
+
+
+/* Local Variables */
+static mss_i2c_status_t status;
+static uint16_t read_value;
+static uint8_t os_bit = 0;
+static uint8_t  rx_buffer[RX_LENGTH];
+static uint16_t hk_adc_avg_volt;
+static uint16_t hk_adc_avg_curr;
+static uint16_t hk_adc_avg_citi_temp;
```

Comment:
These var's are all only used within one function, so they can be made local to these functions. It would also feel safer to me if the variables above them would also be made local to the functions they are used in, there is less risk that if one function gets changed at a later point, it breaks another's assignment to the variable. But if you have no time to address the variables above, then just open up an issue on it and let it be.

Reply:
I made these variables local. But I think the `_avg_` variables should be left global because they are returned outside this file when these fns are called in `msp_exphandler.c`.

Reply:
No, they are not needed. When you return from the function, they are effectively deallocated and the return value stored in a processor register. When the functions are called from `msp_exp_handler.c`, the call allocates the return value from a processor register to the `sprintf` parameter.

But no need to worry about it too much, it is just a minor nice-to-have detail... I'll just open an issue on this and mark this conversation as resolved.